### ISSUE TITLE
Scala option `-Xplugin` doesn't work

### ISSUE THREAD

**Issue by SpriteOvO** [VirtusLab/scala-cli#2621]
**Version(s)**
```
$ scala-cli -version
Scala CLI version: 1.0.5
Scala version (default): 3.3.1
Your Scala CLI version is outdated. The newest version is 1.1.0
It is recommended that you update Scala CLI through the same tool or method you used for its initial installation for avoiding the creation of outdated duplicates.
```

**Describe the bug**

`scala-cli run` doesn't load the specified compiler plugin.

**To Reproduce**

1. Compile a `divbyzero` compiler plugin by `scalac` according to the documentation [here](https://docs.scala-lang.org/overviews/plugins/index.html).

2. `scalac` outputs the expected compilation error.
   ```
   $ scalac -Xplugin:divbyzero.jar test.scala
   test.scala:3: error: definitely division by zero
     val amount = five / 0
                       ^
   one error found
   ```
3. `scala-cli` doesn't output the compilation error, which means the compiler plugin is not loaded.
   ```
   $ scala-cli --scala-version=2.13.12 --scalac-option="-Xplugin:divbyzero.jar" --server=false test.scala
   Exception in thread "main" java.lang.ExceptionInInitializerError
           at Test.main(test.scala)
   Caused by: java.lang.ArithmeticException: / by zero
           at Test$.<clinit>(test.scala:3)
           ... 1 more
   ```
4. More debug information
   ```
   $ scalac -Xplugin:divbyzero.jar -Xshow-phases test.scala | grep divbyzero
        divbyzero  10

   $ scalac -Xplugin:divbyzero.jar -Xplugin-list test.scala | grep divbyzero
   divbyzero - checks for division by zero

   $ scala-cli --scala-version=2.13.12 --scalac-option="-Xplugin:divbyzero.jar" --scalac-option="-Xshow-phases" --server=false test.scala | grep divbyzero
        divbyzero   9

   $ scala-cli --scala-version=2.13.12 --scalac-option="-Xplugin:divbyzero.jar" --scalac-option="-Xplugin-list" --server=false test.scala | grep divbyzero
   [error]  No main class found
   ```

**Expected behaviour**

`scala-cli` should work like `scalac` when used with `-Xplugin`.

**Comment by sequencer** [comment]
Maybe https://github.com/VirtusLab/scala-cli/blob/2dc26d6486576d7882f444eb4d28ccbf06914d05/modules/cli/src/main/scala/scala/cli/commands/shared/SharedOptions.scala#L368-L372 
should resolve jar file here.

**Comment by Gedochao** [comment]
I can replicate the issue... it's weird, as the flag seems to be passed, but the behaviour is inconsistent.
We also have a `--compiler-plugin` option, which seemingly should be handling this, but that doesn't seem to work correctly, either.
I'll try looking at this some more, thanks for reporting

**Comment by Gedochao** [comment]
Okay, so the workaround to make this work correctly is to skip using Bloop with `--server=false`. 
Which is something I kind of suspected from the beginning, but ran into a couple other issues along the way (including i.e. `-Xplugin-list` not working correctly, which is already addressed in #2635)

@SpriteOvO I can see you already tried `--server=false` in the example you have given and it didn't work for you.
I'm pretty sure it does work, let me walk you through the steps (all with Scala CLI, no extra tools required).

I build the custom compiler plugin with the following `DivByZero.scala` source file:

```scala
//> using resourceDir ./resources/
//> using scala  2.13.12

package localhost

import scala.tools.nsc
import nsc.Global
import nsc.Phase
import nsc.plugins.Plugin
import nsc.plugins.PluginComponent

class DivByZero(val global: Global) extends Plugin {
  import global._

  val name = "divbyzero"
  val description = "checks for division by zero"
  val components = List[PluginComponent](Component)

  private object Component extends PluginComponent {
    val global: DivByZero.this.global.type = DivByZero.this.global
    val runsAfter = List[String]("refchecks")
    val phaseName = DivByZero.this.name
    def newPhase(_prev: Phase) = new DivByZeroPhase(_prev)
    class DivByZeroPhase(prev: Phase) extends StdPhase(prev) {
      override def name = DivByZero.this.name
      def apply(unit: CompilationUnit): Unit = {
        for ( tree @ Apply(Select(rcvr, nme.DIV), List(Literal(Constant(0)))) <- unit.body
              if rcvr.tpe <:< definitions.IntClass.tpe)
        {
          global.reporter.error(tree.pos, "definitely divide by zero")
        }
      }
    }
  }
}
```

I keep `scalac-plugin.xml` under `./resources/scalac-plugin.xml`:
```xml
<plugin>
    <name>divbyzero</name>
    <classname>localhost.DivByZero</classname>
</plugin>
```

I build the plugin with Scala CLI:
```bash
scala-cli package DivByZero.scala --power --with-compiler --library -o div-by-zero.jar
# Compiling project (Scala 2.13.12, JVM (17))
# Compiled project (Scala 2.13.12, JVM (17))
# Wrote /Users/pchabelski/IdeaProjects/scala-cli-tests/div-by-zero-2/div-by-zero.jar
```

Just to visualise how the plugin JAR should be structured:
```bash
unzip div-by-zero.jar -d output
# Archive:  div-by-zero.jar
#   inflating: output/META-INF/MANIFEST.MF  
#   inflating: output/localhost/DivByZero.class  
#   inflating: output/localhost/DivByZero$Component$DivByZeroPhase.class  
#   inflating: output/localhost/DivByZero$Component$.class  
#   inflating: output/scalac-plugin.xml  
cd output
tree
# .
# ├── META-INF
# │   └── MANIFEST.MF
# ├── localhost
# │   ├── DivByZero$Component$.class
# │   ├── DivByZero$Component$DivByZeroPhase.class
# │   └── DivByZero.class
# └── scalac-plugin.xml
```


Then make sure it's loaded correctly.
NOTE: `-Xplugin-list` requires passing some (any) input until #2635 is included in the next release.
```bash
scala-cli compile -Xplugin:div-by-zero.jar -Xplugin-list -S 2.13.12 Main.scala --server=false
# divbyzero - checks for division by zero
scala-cli compile -Xplugin:div-by-zero.jar -Xshow-phases -S 2.13|grep divbyzero
#      divbyzero   9
```

I then want to use it with `Main.scala`
```scala
//> using option -Xplugin:div-by-zero.jar
//> using scala 2.13.12

object Main {
  val five = 5
  val amount = five / 0
  def main(args: Array[String]): Unit = {
    println(amount)
  }
}
```
And it does work if I skip Bloop with `--server=false`:
```bash
scala-cli compile -Xplugin:div-by-zero.jar Main.scala --server=false                         
# /Users/pchabelski/IdeaProjects/scala-cli-tests/div-by-zero-2/Main.scala:6: error: definitely divide by zero
#  val amount = five / 0
#                    ^
# 1 error
# Compilation failed
```

It also works with the run command (or no sub-command, which would default to `run`), as in the example from the issue description.
```bash
scala-cli -Xplugin:div-by-zero.jar test.scala --server=false
# /Users/pchabelski/IdeaProjects/scala-cli-tests/div-by-zero-2/test.scala:6: error: definitely divide by zero
#   val amount = five / 0
#                    ^
# 1 error
# Compilation failed
```

Now, it doesn't work correctly with Bloop for some reason, still investigating why.
This is particularly weird, as the compilation passes with Bloop with no error (return code 0), so I'm assuming the plugin isn't working there at all (even though the phase seems to be present), so it might not be the standard "Bloop ate the output" problem.
```bash
scala-cli compile -Xplugin:div-by-zero.jar Main.scala               
# Compiling project (Scala 2.13.12, JVM (17))
# Compiled project (Scala 2.13.12, JVM (17))
```

With Bloop it also breaks same as in the issue description when passed to the `run` sub-command, so I'm assuming something like this was actually being run.
```bash
scala-cli -Xplugin:div-by-zero.jar Main.scala    
# Exception in thread "main" java.lang.ExceptionInInitializerError
#       at test.main(test.scala)
# Caused by: java.lang.ArithmeticException: / by zero
#        at test$.<clinit>(test.scala:6)
#       ... 1 more
```

NOTE: similar replications can be done for Scala 3, refer to the integration test added in #2635 

Also, the `--compiler-plugin` option actually works as expected, but it accepts plugins in coursier dependency format (like `--dep`), while `-Xplugin:<path>` works more like `--extra-jars`. This was surprising to me, but is actually documented correctly (albeit slightly obscure).
I do believe a guide for building and using compiler plugins with Scala CLI is something that should be documented separately.

**Comment by Gedochao** [comment]
TL;DR the workaround is to use `--server=false`. We need compiler plugins to work correctly with Bloop, however, so the issue remains open.

**Comment by SpriteOvO** [comment]
@Gedochao Thanks for your details guide.

Following your instructions (all with scala-cli), `--server=false` worked! But I still can't get `--server=false` to work if they are compiled by extra tools and `scala-cli` just runs them, the compiler plugin is never loaded whether or not the `--server=false` is specified.

The real use case from us is that we use `mill` to build our compiler plugin and packages, then we use `scala-cli` only for testing, so if I understand correctly, there doesn't seem to be a easy way to replace all build system with just `scalc-cli` :/

**Comment by SpriteOvO** [comment]
BTW, the steps in OP is still reproducible (compiles with `scalac`, runs with `scala-cli --server=false`, doesn't load plugin).

I think there won't be any server (e.g. `bloop`) involved in the compilation stage with `scalac`? And then we run these files with `scalc-cli --server=false`. The whole process seems to be without any server involvement, but it doesn't work 🤔

**Comment by Gedochao** [comment]
> Following your instructions (all with scala-cli), `--server=false` worked! But I still can't get `--server=false` to work if they are compiled by extra tools and `scala-cli` just runs them, the compiler plugin is never loaded whether or not the `--server=false` is specified.
> 
> The real use case from us is that we use `mill` to build our compiler plugin and packages, then we use `scala-cli` only for testing, so if I understand correctly, there doesn't seem to be a easy way to replace all build system with just `scalc-cli` :/

Can you double check how the structure inside of the JAR contents produced by `mill` differs from one built with Scala CLI?
I'll check if I can reproduce it when building the plugin with other tools, meanwhile.

> I think there won't be any server (e.g. bloop) involved in the compilation stage with scalac?

Indeed, `scalac` doesn't use Bloop.

**Comment by sequencer** [comment]
It can be reproduced by running `mill lit[2.13.12].run` in https://github.com/chipsalliance/chisel/pull/3596.
llvm-lit is used for replacing jars path from build time, nix is used to provide native dependencies:
`mill lit[2.13.12].run` is alias to `lit out/lit/2.13.12/litConfig.dest`, using `lit out/lit/2.13.12/litConfig.dest -vvv` can see the result:
```
scala-cli --server=false --java-home=/nix/store/93p0sb69hl5j307w40piindi08b44i3p-openjdk-21+35/lib/openjdk --extra-jars=/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.12/scala-library-2.13.12.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.13.12/scala-reflect-2.13.12.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/os-lib_2.13/0.9.1/os-lib_2.13-0.9.1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle_2.13/3.1.0/upickle_2.13-3.1.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/chipsalliance/firtool-resolver_2.13/1.0.0/firtool-resolver_2.13-1.0.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/json4s/json4s-native_2.13/4.0.6/json4s-native_2.13-4.0.6.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/github/alexarchambault/data-class_2.13/0.2.5/data-class_2.13-0.2.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-text/1.10.0/commons-text-1.10.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/github/scopt/scopt_2.13/3.7.1/scopt_2.13-3.7.1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/geny_2.13/1.0.0/geny_2.13-1.0.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/ujson_2.13/3.1.0/ujson_2.13-3.1.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upack_2.13/3.1.0/upack_2.13-3.1.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle-implicits_2.13/3.1.0/upickle-implicits_2.13-3.1.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/dev/dirs/directories/26/directories-26.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/outr/scribe_2.13/3.11.5/scribe_2.13-3.11.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier_2.13/2.1.5/coursier_2.13-2.1.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/json4s/json4s-core_2.13/4.0.6/json4s-core_2.13-4.0.6.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/json4s/json4s-native-core_2.13/4.0.6/json4s-native-core_2.13-4.0.6.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-lang3/3.12.0/commons-lang3-3.12.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/upickle-core_2.13/3.1.0/upickle-core_2.13-3.1.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/outr/perfolation_2.13/1.2.9/perfolation_2.13-1.2.9.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/sourcecode_2.13/0.3.0/sourcecode_2.13-0.3.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-collection-compat_2.13/2.11.0/scala-collection-compat_2.13-2.11.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/outr/moduload_2.13/1.1.6/moduload_2.13-1.1.6.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/github/plokhotnyuk/jsoniter-scala/jsoniter-scala-core_2.13/2.13.5.2/jsoniter-scala-core_2.13-2.13.5.2.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier-core_2.13/2.1.5/coursier-core_2.13-2.1.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier-cache_2.13/2.1.5/coursier-cache_2.13-2.1.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier-proxy-setup/2.1.5/coursier-proxy-setup-2.1.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/json4s/json4s-ast_2.13/4.0.6/json4s-ast_2.13-4.0.6.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/json4s/json4s-scalap_2.13/4.0.6/json4s-scalap_2.13-4.0.6.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/thoughtworks/paranamer/paranamer/2.8/paranamer-2.8.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/github/alexarchambault/concurrent-reference-hash-map/1.1.0/concurrent-reference-hash-map-1.1.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.13/2.1.0/scala-xml_2.13-2.1.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/coursier-util_2.13/2.1.5/coursier-util_2.13-2.1.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/get-coursier/jniutils/windows-jni-utils/0.3.3/windows-jni-utils-0.3.3.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/codehaus/plexus/plexus-archiver/4.7.1/plexus-archiver-4.7.1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/codehaus/plexus/plexus-container-default/2.1.1/plexus-container-default-2.1.1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/virtuslab/scala-cli/config_2.13/0.2.1/config_2.13-0.2.1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/io/github/alexarchambault/windows-ansi/windows-ansi/0.0.5/windows-ansi-0.0.5.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/javax/inject/javax.inject/1/javax.inject-1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/codehaus/plexus/plexus-utils/3.5.1/plexus-utils-3.5.1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/codehaus/plexus/plexus-io/3.4.1/plexus-io-3.4.1.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/commons-io/commons-io/2.11.0/commons-io-2.11.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/apache/commons/commons-compress/1.23.0/commons-compress-1.23.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/slf4j/slf4j-api/1.7.36/slf4j-api-1.7.36.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/iq80/snappy/snappy/0.4/snappy-0.4.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/tukaani/xz/1.9/xz-1.9.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/com/github/luben/zstd-jni/1.5.5-2/zstd-jni-1.5.5-2.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/codehaus/plexus/plexus-classworlds/2.6.0/plexus-classworlds-2.6.0.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/apache/xbean/xbean-reflect/3.7/xbean-reflect-3.7.jar:/home/sequencer/.cache/coursier/v1/https/repo1.maven.org/maven2/org/fusesource/jansi/jansi/1.18/jansi-1.18.jar:/home/sequencer/projects/chisel/compile-resources:/home/sequencer/projects/chisel/src/main/resources:/home/sequencer/projects/chisel/out/chisel/2.13.12/compile.dest/classes:/home/sequencer/projects/chisel/svsim/compile-resources:/home/sequencer/projects/chisel/svsim/src/main/resources:/home/sequencer/projects/chisel/out/svsim/2.13.12/compile.dest/classes:/home/sequencer/projects/chisel/firrtl/compile-resources:/home/sequencer/projects/chisel/firrtl/src/main/resources:/home/sequencer/projects/chisel/out/firrtl/2.13.12/compile.dest/classes:/home/sequencer/projects/chisel/macros/compile-resources:/home/sequencer/projects/chisel/macros/src/main/resources:/home/sequencer/projects/chisel/out/macros/2.13.12/compile.dest/classes:/home/sequencer/projects/chisel/core/compile-resources:/home/sequencer/projects/chisel/core/src/main/resources:/home/sequencer/projects/chisel/out/core/2.13.12/compile.dest/classes:/home/sequencer/projects/chisel/binder/compile-resources:/home/sequencer/projects/chisel/binder/src/main/resources:/home/sequencer/projects/chisel/out/circtpanamabinder/2.13.12/compile.dest/classes:/home/sequencer/projects/chisel/filecheckutility/utility/compile-resources:/home/sequencer/projects/chisel/filecheckutility/utility/resources:/home/sequencer/projects/chisel/out/filecheckutility/2.13.12/compile.dest/classes --scala-version=2.13.12 --scala-option="-Xplugin:/home/sequencer/projects/chisel/out/plugin/2.13.12/jar.dest/out.jar" --java-opt="--enable-native-access=ALL-UNNAMED --enable-preview -Djava.library.path=/nix/store/0mp3jx5b444hy32n878mcx9sqv48i25y-circt-all/lib" /home/sequencer/projects/chisel/lit/tests/CompilerPlugin/WithCompilerPlugin.sc
```

We use `--server=false --scala-option="-Xplugin:/home/sequencer/projects/chisel/out/plugin/2.13.12/jar.dest/out.jar"`, but it complains: 
`Exception in thread "main" java.lang.AssertionError: assertion failed: The Chisel compiler plugin is now required for compiling Chisel code. Please see https://github.com/chipsalliance/chisel3#build-your-own-chisel-projects.
`.

**Comment by sequencer** [comment]
@Gedochao does #2654 resolve this issue?

**Comment by Gedochao** [comment]
> @Gedochao does #2654 resolve this issue?

@sequencer Nope, it rather fixes a separate bug which has been affecting the debugging of this one. They're indirectly connected.
I'm still looking into a solution for this.

**Comment by sequencer** [comment]
> @sequencer Nope, it rather fixes a separate bug which has been affecting the debugging of this one. They're indirectly connected.
I'm still looking into a solution for this.

Thanks, really appreciate your work!

**Comment by Gedochao** [comment]
@sequencer @SpriteOvO Can you confirm for me, if the issue is still replicable for you with the `--server=false` flag after bumping to Scala CLI `v1.1.2`? 
If you notice any change in behaviour/output, I'd also like to know.
I am aware the problem is still present for Bloop, but I believe it should now work correctly with `--server=false`.

**Comment by SpriteOvO** [comment]
@Gedochao I can confirm that the minimum example is no longer reproducible in `v1.1.2`.

```
> ./scala-cli-v1.1.1 --scala-version=2.13.12 --scalac-option="-Xplugin:divbyzero.jar" --server=false Test.scala
Exception in thread "main" java.lang.ExceptionInInitializerError
        at Test.main(Test.scala)
Caused by: java.lang.ArithmeticException: / by zero
        at Test$.<clinit>(Test.scala:3)
        ... 1 more

> ./scala-cli-v1.1.2 --scala-version=2.13.12 --scalac-option="-Xplugin:divbyzero.jar" --server=false Test.scala
~/scala-plugin-test/Test.scala:3: error: definitely division by zero
  val amount = five / 0
                    ^
1 error
Compilation failed
```

We will test it on our real project and give feedback later. Much appreciate your work! 👍

**Comment by SpriteOvO** [comment]
It works like a charm in our project now! (chipsalliance/chisel#3596)

Thanks again! @Gedochao
